tls/s2n_x509_validator.c

```diff
@@ -213,95 +214,156 @@ int s2n_x509_validator_set_max_chain_depth(struct s2n_x509_validator *validator,
+
+        return S2N_RESULT_OK;
+    }
+
+    if (current_name->type == GEN_IPADD) {
+        *san_found = true;
+
+        /* try to validate an IP address if it's in the subject alt name. */
+        const unsigned char *ip_addr = current_name->d.iPAddress->data;
+        size_t ip_addr_len = (size_t) current_name->d.iPAddress->length;
+
+        RESULT_STACK_BLOB(address, INET6_ADDRSTRLEN + 1, INET6_ADDRSTRLEN + 1);
+
+        if (ip_addr_len == 4) {
+            RESULT_GUARD(s2n_inet_ntop(AF_INET, ip_addr, &address));
+        } else if (ip_addr_len == 16) {
+            RESULT_GUARD(s2n_inet_ntop(AF_INET6, ip_addr, &address));
+        } else {
+            /* we aren't able to parse this value so skip it */
+            return S2N_RESULT_ERROR;
+        }
+
+        /* strlen should be safe here since we made sure we were null terminated AND that inet_ntop succeeded */
+        const char *name = (const char *) address.data;
+        size_t name_len = strlen(name);
+
+        RESULT_ENSURE(conn->verify_host_fn(name, name_len, conn->data_for_verify_host), S2N_ERR_CERT_UNTRUSTED);
+
+        return S2N_RESULT_OK;
+    }
+
+    /* we don't understand this entry type so skip it */
+    return S2N_RESULT_ERROR;
+}
+
+static S2N_RESULT s2n_verify_host_information_san(struct s2n_connection *conn, X509 *public_cert, bool *san_found)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(public_cert);
+    RESULT_ENSURE_REF(san_found);
+
+    *san_found = false;
+
+    DEFER_CLEANUP(STACK_OF(GENERAL_NAME) *names_list = NULL, GENERAL_NAMES_free_pointer);
+
+    names_list = X509_get_ext_d2i(public_cert, NID_subject_alt_name, NULL, NULL);
+
+    int n = sk_GENERAL_NAME_num(names_list);
+    for (int i = 0; i < n; i++) {
+        GENERAL_NAME *current_name = sk_GENERAL_NAME_value(names_list, i);
+
+        /* if we get any valid entries then return */
+        if (s2n_result_is_ok(s2n_verify_host_information_san_entry(conn, current_name, san_found))) {
+            return S2N_RESULT_OK;
+        }
+    }
+
+    /* if an error was set, then don't override it */
+    if (s2n_errno) {
+        return S2N_RESULT_ERROR;
+    }
+
+    RESULT_BAIL(S2N_ERR_CERT_UNTRUSTED);
+}
+
+static S2N_RESULT s2n_verify_host_information_common_name(struct s2n_connection *conn, X509 *public_cert, bool *cn_found)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(public_cert);
+    RESULT_ENSURE_REF(cn_found);
+
+    X509_NAME *subject_name = X509_get_subject_name(public_cert);
+    RESULT_ENSURE(subject_name, S2N_ERR_CERT_UNTRUSTED);
+
+    int next_idx = 0, curr_idx = -1;
+    while ((next_idx = X509_NAME_get_index_by_NID(subject_name, NID_commonName, curr_idx)) >= 0) {
+        curr_idx = next_idx;
+    }
+
+    RESULT_ENSURE(curr_idx >= 0, S2N_ERR_CERT_UNTRUSTED);
+
+    ASN1_STRING *common_name = X509_NAME_ENTRY_get_data(X509_NAME_get_entry(subject_name, curr_idx));
```

Comment:
I think X509_NAME_get_entry can return NULL so I don't like the idea of daisy-chaining it with X509_NAME_ENTRY_get_data.

Reply:
I had to revert my fix here because openssl changes constness between different versions, unfortunately. It'll have to be a refactor for another day 😄